File: core/models/sectioning_strategy.py
from abc import ABC, abstractmethod
from core.models.note import Note
from core.models.section import Section

class SectioningStrategy(ABC):
    """ Abstract class for sectioning strategy """

    @abstractmethod
    def process(self, note: Note) -> list[Section]:
        """ Section note based on the desired number of section """
        pass

class StaticSectioningStrategy(SectioningStrategy):
    """ Class for static sectioning strategy """

    def __init__(self, num_of_section):
        self.num_of_section = num_of_section

    def process(self, note: Note) -> list[Section]:
        """
            For static sectioning, simply section by number of lines to achieve
            desired number of sections
        """

        lines = [line.strip() for line in note.content.split("\n")]  # Remove trailing newlines

        num_of_section = self.num_of_section

        total_lines = len(lines)
        section_size = max(1, total_lines // num_of_section)  # Ensure at least 1 line per section
        sections = []
        for i in range(num_of_section):
            start = i * section_size
            end = (i + 1) * section_size if i < num_of_section - 1 else total_lines  # Last section gets remaining lines
            section_text = "\n".join(lines[start:end])  # Concatenate lines into one string
            sections.append(Section(section_text))

        return sections



        

File: src/lib/models/sectioning-strategy.ts
import { Note } from "./note";
import { Section } from "./section";

/**
 * Abstract class for sectioning strategy
 */
export abstract class SectioningStrategy {
  /**
   * Section note based on the desired number of sections
   */
  abstract process(note: Note): Section[];
}

/**
 * Class for basic sectioning strategy
 * Sends the entire note to the LLM in one go
 */
export class BasicSectioningStrategy extends SectioningStrategy {
  /**
   * For basic sectioning, return the entire note as a single section
   */
  process(note: Note): Section[] {
    return [new Section(note.content)];
  }
}

/**
 * Class for static sectioning strategy
 */
export class StaticSectioningStrategy extends SectioningStrategy {
  private numOfSection: number;

  constructor(numOfSection: number) {
    super();
    this.numOfSection = numOfSection;
  }

  /**
   * For static sectioning, simply section by number of lines to achieve
   * desired number of sections
   */
  process(note: Note): Section[] {
    // Remove trailing newlines
    const lines = note.content.split("\n").map((line) => line.trim());

    const numOfSection = this.numOfSection;
    const totalLines = lines.length;

    // Ensure at least 1 line per section
    const sectionSize = Math.max(1, Math.floor(totalLines / numOfSection));
    const sections: Section[] = [];

    for (let i = 0; i < numOfSection; i++) {
      const start = i * sectionSize;
      // Last section gets remaining lines
      const end = i < numOfSection - 1 ? (i + 1) * sectionSize : totalLines;
      const sectionText = lines.slice(start, end).join("\n");
      sections.push(new Section(sectionText));
    }

    return sections;
  }
}
